pinot-core/src/main/java/org/apache/pinot/core/query/reduce/AggregationDataTableReducer.java

```diff
@@ -41,33 +45,46 @@
 public class AggregationDataTableReducer implements DataTableReducer {
 
   private final AggregationFunction[] _aggregationFunctions;
+  private final List<AggregationInfo> _aggregationInfos;
+  private final int _numAggregationFunctions;
   private final boolean _preserveType;
+  private boolean _responseFormatSql;
 
   AggregationDataTableReducer(BrokerRequest brokerRequest, AggregationFunction[] aggregationFunctions,
       QueryOptions queryOptions) {
     _aggregationFunctions = aggregationFunctions;
+    _aggregationInfos = brokerRequest.getAggregationsInfo();
+    _numAggregationFunctions = aggregationFunctions.length;
     _preserveType = queryOptions.isPreserveType();
+    _responseFormatSql = queryOptions.isResponseFormatSQL();
   }
 
   /**
-   * Reduces data tables and sets aggregations results into BrokerResponseNative::AggregationResults
+   * Reduces data tables and sets aggregations results into
+   * 1. ResultTable if _responseFormatSql is true
+   * 2. AggregationResults by default
    */
   @Override
-  public void reduceAndSetResults(String tableName, DataSchema dataSchema, Map<ServerRoutingInstance, DataTable> dataTableMap,
-      BrokerResponseNative brokerResponseNative, BrokerMetrics brokerMetrics) {
+  public void reduceAndSetResults(String tableName, DataSchema dataSchema,
+      Map<ServerRoutingInstance, DataTable> dataTableMap, BrokerResponseNative brokerResponseNative,
+      BrokerMetrics brokerMetrics) {
+
     if (dataTableMap.isEmpty()) {
+      if (_responseFormatSql) {
```

Comment:
Any way to avoid this `if` check in various places? I guess the issue is that we are setting different fields in BrokerResponse, or else, we could have just used an interface over ResultTable and the old format?

Reply:
I tried to add another layer of abstraction for sql vs pql. The group by code got very messy. There's a lot of code that needs to be shared across the two. 
I can try it again, but would prefer to revisit after this much is merged. That'll make it easier to code and also review